rclrs/src/node.rs

```diff
@@ -360,6 +360,84 @@ impl Node {
         domain_id
     }
 
+    /// Creates a `ParameterBuilder` that can be used to set parameter declaration options and
+    /// declare a parameter as `Optional`, `Mandatory` or `ReadOnly`.
+    ///
+    /// # Example
+    /// ```
+    /// # use rclrs::{Context, ParameterRange, RclrsError};
+    /// let context = Context::new([])?;
+    /// let node = rclrs::create_node(&context, "domain_id_node")?;
+    /// // Set it to a range of 0-100, with a step of 2
+    /// let range = ParameterRange {
+    ///     lower: Some(-100),
+    ///     upper: Some(100),
+    ///     step: Some(2),
+    /// };
+    /// let param = node.declare_parameter("int_param", 10)
+    ///                 .range(range)
+    ///                 .mandatory()
+    ///                 .unwrap();
+    /// assert_eq!(param.get(), 10);
+    /// param.set(50).unwrap();
+    /// assert_eq!(param.get(), 50);
+    /// // Out of range, will return an error
+    /// assert!(param.set(200).is_err());
+    /// # Ok::<(), RclrsError>(())
+    /// ```
+    pub fn declare_parameter<T: Declarable>(
+        &self,
+        name: &str,
+        default_value: T,
```

Comment:
I think there should be a way to not pass a default value. Ideally, there would also be a way to declare a parameter with a _fixed_ value, regardless of what override was given (in `rclcpp`, it's the `ignore_overrides` parameter, which is a somewhat ugly way to solve it if you ask me).

So, if we omit the default value from this function and instead add a `default()` function to the parameter builder, then it should be a failure if no override was provided and the default was not set.

Reply:
> I think there should be a way to not pass a default value.

That exists for Optional parameters.

For it to exist for Mandatory parameters we would need it to return an error if no parameter override value is found. That should be pretty straightforward by adding another error variant to `DeclarationError`.

> Ideally, there would also be a way to declare a parameter with a fixed value, regardless of what override was given

It should be pretty easy to add this option to the parameter builder.

Reply:
Alternatively, `default_value` could become a field set by the builder and we can get rid of `declare_optional_parameter`. If the user selects `.mandatory()` or `.read_only()` when neither a default nor override value is set then they get back an error when they build.

My only hesitation with this approach is in not sure if the Rust compiler can infer the parameter type if the default value is provided through a method on the builder. I have a personal distaste for turbofish syntax so I'd like to avoid forcing that on the user if possible.

Does this sound like something worth testing out @luca-della-vedova ?

Reply:
Agreed on converging all those different signatures into new builder methods.

Do you have a branch where you're experimenting with this? I might be able to take a look and get the inference working.

Reply:
I'll admit not really having a clear idea about what I'm doing since I'm a bit confused by the bounds needed for these functions, but I committed the current refactoring work https://github.com/ros2-rust/ros2_rust/pull/332/commits/be4b977d890f1c6a046c748d9b509d5c5b0f1be8